### ISSUE TITLE
Lines vs Events

### ISSUE THREAD

**Issue by quisquous** [OverlayPlugin/OverlayPlugin#81]
Before we add too many new line types, I wanted to have a brief discussion and solicit feelings about guidelines for adding new lines.  Ravahn sent some guidance that they were a little concerned about adding lines that were too noisy and would create additional log processing delays for people who are already having this issue.  Obviously it's multidimensionally subjective what "too many" log lines is based on existing logs, content, people's computers, but I think it's something to keep in mind.

In my mind, OverlayPlugin now has two methods for providing data: lines and events.  Network lines are an event itself, but they behave differently in that lines are stored on disk and almost every overlay subscribes to the log event.  Only people running an overlay that subscribes to other events will get that data.  I would also argue that there's probably more processing per-line than per-event just based on the fact that there are more implicit subscribers to lines than to events.  Lines are also a way to get data to Triggernometry or fflogs in a way that events are not.

In my mind, here's what I imagine as a criteria for lines vs events:

* Lines: data for triggers, data used for log analysis/imports/playback
* Events: things that are more ephemeral, data for UI (e.g. job stuff, enmity stuff)

To apply this to various hypothetical data sources:
* Custom line registration: line, because needed for importing/analysis
* Map events: line, because needed for triggers
* Fates/CEs: event?, not really something you would write a trigger for, ephemeral data, not really needed for analysis/playback
* Fishing !/!!/!!! headmarkers: line, not combat data but is something people would want to write triggers for
* A8S/P1S weapon attachment data: line, because needed for triggers
* Start/Stop combat: line, needed when importing to split encounters properly
* Targeting changes: event, since ephemeral and not really needed for triggers?

Am I overthinking this? Do other people have other heuristics they'd rather apply? Does anybody have a suggestion of what a "% of log lines" we should be trying to aim for or be concerned about with new log line types?  I certainly think that probably we are at a point where we will add a bunch of lines in the short term but it will slow down after that.  However, I mostly want to come to some consensus before adding too many.

**Comment by quisquous** [comment]
Really the heart of the question for me is to separate out "is this data worth logging to disk, and having all of our log-processing machinery chew through it" vs "is this data just something that we want OverlayPlugin to provide in some form or another".

I think this is my bias, but my preference also is to have any log lines OverlayPlugin provides to always be in the log file and to not be configurable or optional.  I do enough tech support already for people who have chat lines off or various other configuration weirdness to not want to have even more knobs to have to go look at to understand why something is broken or missing from a log file.

**Comment by xpdota** [comment]
Short version for me is: I don't care, I can use either (or in some cases both, so that both log replaying and live work equally well, since the event can be cached). Since I can record the entire Websocket input, I can also capture whatever stuff doesn't have a log line.

Things like targeting changes *can* be useful for triggers (think tank swaps). I think if you really stretch it, just about anything *could* be useful as a line. Rather than choosing "debate what is useful enough to be in the log" as the long-term solution, do we instead just standardize a "saved websocket session" format that includes everything? There's already issues of live vs replay reproducibility because of things like combatant positions not always accurately being reflected in the log at the time of a cast, so saving the entire socket flow is the only way to get *existing* functionality fully working when replaying anyway.

**Comment by valarnin** [comment]
Sort of off on a tangent here, but re: performance.

I'd really like to get some profiler data from someone with a weaker machine to see what's slowing down log line processing the most.

I've never had issues with log line processing, even in world fates with 300+ people and `Log all Network Packets` enabled, with my i7 8700k.

I can only assume this is some issue C#-side with processing the lines, or at the HDD level with writing to/reading from the file, as anything overlay-side regardless of the point of processing is happening in another thread which would never trigger the "log processing running behind" error message.

If it's C#-side, it would fall into one of two camps:
1. ACT itself/FFXIV_ACT_Plugin - Nothing we can do about it
2. OverlayPlugin/Cactbot/Triggernometry/some other open-source project - Improve performance to help alleviate issues

---

Re: the original thrust of this issue. I guess the question from me is that, are there cases where we as devs will want e.g. Fate/CE lines, or other lines in the future, which won't really be needed by end-users?

> - Fishing !/!!/!!! headmarkers: line, not combat data but is something people would want to write triggers for

This is a curious example because I believe that we could currently support "lines" like these for triggers without ever writing them to file, by invoking <https://advancedcombattracker.com/apidoc/html/M_Advanced_Combat_Tracker_FormActMain_ParseRawLogLine.htm>, which is maybe a whole other discussion.

Should we allow the registering caller to indicate via parameter if the event type should be "write to disk" vs "emit as line only" vs "fire an event"?

**Comment by panicstevenson** [comment]
Personally, I haven't seen the ACT is having trouble parsing logs message in a very long time; I don't think it was related to any hardware changes, but maybe. I also had the initial thought in the same vein as valarnin in that "can logs just be processed faster (or in parallel)?" Although that skirts the true design issue here for the sake of a bandaid (albeit maybe effective enough).

I think as far as the other things listed here -- I don't think any existing logs can really be removed without causing some detriment (if that were even an available option), which leads into the same situation on the other side once additional logs have been added -- they're hard to take away once added.

As for some heuristic to what's okay to put in and what isn't, I really couldn't say. Likely you want things that cause triggers to be available within logs for the sake of debugging, changing, etc. (extending down to base ACT, as far as I'm concerned). We could say that MapEvents are okay and necessary given the evolution of fight mechanics, but comparing a modern fight to one from A Realm Reborn will have very different logging amounts for things like that, making it relatively difficult to gate what is and isn't okay based off volume, leaving us with (as far as I can see) two metrics: value added and accessibility.

I consider the former to be the more obvious: How nice would it be to have a trigger? And the latter to be how nice it would be to have just that log line without supplemental logic. I don't know the full state of the third party ecosystem (I think one point of this post), but if everyone is capable of using events and a log line is not useful at a fundamental level, it may be worth not logging. However, at that point, then debugging will likely be more difficult without some way for all plugins to capture and preserve and output these events.

To the parallelism point in parentheses above, I think things can be logged into some sidecar log file, but I (again) don't know if that addresses the points made in quisquous' second post. Anyway, this is all to say I think it's somewhat nebulous and unless the tooling gets a lot better for the average user to create support files for debugging I think most things will default to being logged. Which then leads to, maybe implement the least amount upfront and try to peel off things one at a time. In this same vein, I would also say anything that grows sublinearly with log amounts is probably also "okay" here; for example, if there were a log line added at the start of every instance to calculate the random offset for items and head markers, since that could then be used in the logic of other callouts (or maybe eventually just fix all of the head markers in their log lines -- going back to the accessibility point above).

**Comment by xpdota** [comment]
One counterpoint - if we *were* to have something like one line to give us the headmarker offset, or something like RSVs, there is value in having that be available as an event rather than a log line (pretend it's not feasible to just fix up the headmarker/ability lines with the corrected offset/RSV values), since an event supports clients that connect after the zone has started. That's one of the biggest advantages of events over log lines IMO (or at the very least, is a reason to do both in those circumstances, like for map/zone/party changes).

Also, something that I don't think is talked about is the sheer variance in number of log lines simply due to a user's battle log settings (or whether they have the "hide chat log for privacy" setting enabled). A simple cast/snapshot/resolve goes from 3 lines to 6 if the relevant battle log options are enabled. I think it's probably fine to tell someone "hey, if you're seeing your parsing lag behind, turn off unused battle log stuff as a workaround". You've also got some plugins dumping their debug log as fake echo chat lines, which would have a disproportionate impact on performance because any regex/parsing looking for echo lines will have to process further into the line before throwing it out.

**Comment by valarnin** [comment]
Re: log lines for things that only happen on zone-in such as RSV or headmarker offset info, I don't think that's feasible anyways because it'll break if someone doesn't have ACT open when zoning in, or has to restart ACT mid-zone. All triggers should be written as if ACT were started 15 seconds before the pull, because that's the case fairly often.

**Comment by xpdota** [comment]
Someone might have ACT open, but not have the overlay/client connected yet (think remote cactbot), perhaps the overlay itself just hasn't fully connected yet, or just someone restarting the WS client for dev/debugging purposes. Same idea as zone/map/party change - these work fine if the overlay connects after the lines have been emitted because OP can cache it. At the same time, they're driven by log lines in the first place, so in most of these cases, the answer is "do both options". I'd advocate for that, since it still works if you open a plain log file.

But I still think it is a good idea to have an "enhanced log" format that captures websocket data rather than just log lines. As I mentioned earlier, some triggers require accurate combatants data to function correctly (even if the log line itself has stale data), meaning you need a capture of that data for them to be properly testable. I do this anyway because sometimes there's even more data sources needed for an accurate reconstruction (e.g. I currently need Telesto for a properly sorted party list).

**Comment by valarnin** [comment]
I have some thoughts with regards to more accurate combatant data going forward but that's not really directly relevant to this discussion.

Circling back around to the "how many lines is too many" question, I think I can summarize my thoughts as:

1. Extra lines during a time where lines aren't really heavy anyways (e.g. fate/CE lines in overworld zones) should be fine as long as they're not extremely spammy
2. Extra lines during heavy combat (24-man content or 48-man content) should be heavily scrutinized since those are the most likely to impact performance
3. Extra lines during normal 8-man content should be fine as long as they aren't excessive

As far as the question of "line vs event", I think that anything generated based off of network data should be a line. Even if it's spammy, that spam can be reduced/eliminated by C#-side processing before emitting the line. There's no point in making it an event for queue-ing purposes if there's always a chance that ACT wasn't open when the line arrived to begin with, that will just lead to inconsistent behavior.

Note that map change/player change/misc other info as mentioned elsewhere are detected via memory by FFXIV_ACT_Plugin, not via network data, that's why they always get emitted when ACT is started.

**Comment by xpdota** [comment]
> There's no point in making it an event for queue-ing purposes if there's always a chance that ACT wasn't open when the line arrived to begin with, that will just lead to inconsistent behavior.

It might lead to inconsistent behavior, but better to generally have it than not as long as nothing is dependent on it for anything more than bells and whistles. e.g. before ChangeMap got a cached event to go along with the log line, the only impact to me was not seeing the map background image. It's still better to have it cached than not cached, since it's only used to improve a UI view.

But again, I don't think these are necessarily mutually exclusive. It's really more of a 3-way choice - line, event, or both. Generally, I think that when you factor in caching behavior, any client that connects to the WS should be able to reconstruct state, just as if it had been connected the instant ACT started up. If ACT itself missed the line, then you wouldn't have that data either way.

**Comment by quisquous** [comment]
> Circling back around to the "how many lines is too many" question, I think I can summarize my thoughts as:
>
> Extra lines during a time where lines aren't really heavy anyways (e.g. fate/CE lines in overworld zones) should be fine as long as they're not extremely spammy
> Extra lines during heavy combat (24-man content or 48-man content) should be heavily scrutinized since those are the most likely to impact performance
> Extra lines during normal 8-man content should be fine as long as they aren't excessive

This makes sense to me as a guideline.  I am sorry to put up such a nebulous discussion here.  I know there's not really great answers.  I just wanted to see where thoughts were.

> As far as the question of "line vs event", I think that anything generated based off of network data should be a line. Even if it's spammy, that spam can be reduced/eliminated by C#-side processing before emitting the line. There's no point in making it an event for queue-ing purposes if there's always a chance that ACT wasn't open when the line arrived to begin with, that will just lead to inconsistent behavior.

I think this is somewhat unclear, as the line (pun not intended) between network and memory data is murky.  Memory data is there because network data has been sent, and some code has executed, changing memory.  So, memory data is just indirect network data.  It may be in an easier form or easier to find, but it's all "data from the server" in my opinion.

In terms of ACT missing lines due to startup, I would maybe suggest that we do something similar to what I did for cactbot's log splitter.  I found that you could split lines into three categories: (1) lines that always need to be included (startup versions, etc), (2) lines that you only need the last instance of (zone changing, player name, party), (3) lines that you don't need to be cached at all.  We could have an explicit "give me the cached lines" request.  (Maybe rsv lines create a 4th category of "lines to send once per zone, clear when zoning")  We could also just send these cached lines immediately when a client connects as the first lines they receive.  That way we wouldn't need to add an event that corresponds with lines just so it can be requested.

**Comment by quisquous** [comment]
> Should we allow the registering caller to indicate via parameter if the event type should be "write to disk" vs "emit as line only" vs "fire an event"?

Personally, I would like to remove as much optionality in this as we can.  I think the logic I'd like to follow is "OverlayPlugin provides X Y Z lines" -> "if OverlayPlugin is enabled, X Y Z lines will always be present in the log file".  That way there's no worry about "is X line firing? I don't know, it doesn't get written to the file" or "is X line firing? well let me check these options about when it fires or not to see if it's a bug or you're just in a weird setup".  I'd like to make this as simple as possible, configuration-wise.

**Comment by valarnin** [comment]
> This makes sense to me as a guideline. I am sorry to put up such a nebulous discussion here. I know there's not really great answers. I just wanted to see where thoughts were.

I think the discussion needed to be had, and I'd rather it be in a public place where it can be reviewed easily later, rather than getting lost in Discord or something. I'd also like to see some better-worded/more concrete guidelines documented somewhere as well when we've settled on things.

> Memory data is there because network data has been sent, and some code has executed, changing memory.

I think for me the distinction here is "network data is realtime, memory data is polled at an interval or on-demand". I was trying to convey two separate points and didn't state them explicitly enough, so let me provide my thoughts on them directly:
1. Network data - Always a log line.
2. Memory data - Could be a log line, an event (e.g. getCombatants), or a subscription (e.g. EnmityTargetData)

> I found that you could split lines into three categories: (1) lines that always need to be included (startup versions, etc), (2) lines that you only need the last instance of (zone changing, player name, party), (3) lines that you don't need to be cached at all.

I think that these categories make sense.

> We could have an explicit "give me the cached lines" request. (Maybe rsv lines create a 4th category of "lines to send once per zone, clear when zoning") We could also just send these cached lines immediately when a client connects as the first lines they receive. That way we wouldn't need to add an event that corresponds with lines just so it can be requested.

I could go either way on this. I'd be concerned with option 2 (send on connect) that we're forcing overlays that want that information to essentially duplicate the logic that we're already discussing adding here though with regards to caching.

> Personally, I would like to remove as much optionality in this as we can. I think the logic I'd like to follow is "OverlayPlugin provides X Y Z lines" -> "if OverlayPlugin is enabled, X Y Z lines will always be present in the log file".

I think that maybe there's a miscommunication here on my end. I was saying, essentially:

1. OverlayPlugin-provided lines should always go in the log file.
3. Downstream ACT plugins that interface with OverlayPlugin and leverage the line registration function, e.g. potentially cactbot, or triggernometry, etc, should be able to indicate if they want the log lines to be written to file or not.

**Comment by quisquous** [comment]
> I'd also like to see some better-worded/more concrete guidelines documented somewhere as well when we've settled on things.

:+1:

> I think for me the distinction here is "network data is realtime, memory data is polled at an interval or on-demand". I was trying to convey two separate points and didn't state them explicitly enough, so let me provide my thoughts on them directly:

OH! This makes a lot more sense.  Thanks for the explanation.  I think I mooostly agree.  I think there could be some things that are too noisy that you'd want to poll even if it was network data (e.g. like ... if we tracked where combatants were from network data instead of memory, we probably wouldn't push every combatant movement into the log for noise reasons?), but that's probably me being pedantic here.

> I could go either way on this. I'd be concerned with option 2 (send on connect) that we're forcing overlays that want that information to essentially duplicate the logic that we're already discussing adding here though with regards to caching.

I'm not sure I understand the concern here.  My view is that it would let overlays use only log lines rather than also need to use events for these sorts of things.  I'm not sure what you mean by duplicating logic, as I was imagining that an overlay's logic to handle the log lines sent on connect or sent when they occur would be the same path.

> Downstream ACT plugins that interface with OverlayPlugin and leverage the line registration function, e.g. potentially cactbot, or triggernometry, etc, should be able to indicate if they want the log lines to be written to file or not.

That seems reasonable to me if this is something that other plugins want.  Definitely agreed that OverlayPlugin lines should be unconditional.  Personally, I don't see myself adding lines to cactbot.  I would probably promote everything into OverlayPlugin if I could, although that just gets back to the "should cactbot plugin merge into OverlayPlugin" discussion a little.

**Comment by xpdota** [comment]
Welcome back!

A few of my thoughts:

* I would also like for the log file to be a more complete capture. Currently, I have my own save format, but it's not interoperable with anything else, so it would be nice to have a single unified format.
* I agree with the overhead part - one reason I'd like to move away from my current save format is because the JSON version of these events vs the raw log line is a bit bloated:

```
21|2023-03-17T17:05:22.0490000-07:00|400117AD|Bunshin|4405|Spinning Edge|40011651|Omega|710003|104E0000|53D|9F8000|1B|44058000|0|0|0|0|0|0|0|0|0|0|5661608|8557964|10000|10000|||100.08|100.11|0.00|0.80|68561|71428|10000|10000|||87.72|99.29|0.00|1.50|0000F325|0|1|62dcc32cf1be3b26
```
vs
```
{"type":"LogLine","line":["21","2023-03-17T17:05:22.0490000-07:00","400117AD","Bunshin","4405","Spinning Edge","40011651","Omega","710003","104E0000","53D","9F8000","1B","44058000","0","0","0","0","0","0","0","0","0","0","5661608","8557964","10000","10000","","","100.08","100.11","0.00","0.80","68561","71428","10000","10000","","","87.72","99.29","0.00","1.50","0000F325","0","1","62dcc32cf1be3b26"],"rawLine":"21|2023-03-17T17:05:22.0490000-07:00|400117AD|Bunshin|4405|Spinning Edge|40011651|Omega|710003|104E0000|53D|9F8000|1B|44058000|0|0|0|0|0|0|0|0|0|0|5661608|8557964|10000|10000|||100.08|100.11|0.00|0.80|68561|71428|10000|10000|||87.72|99.29|0.00|1.50|0000F325|0|1|62dcc32cf1be3b26"}
```
(I feel like the splitting should just happen client-side if it is actually needed - no way it's less overhead to split on the C# side, serialize, and deserialize them vs just having the split happen on the client side, but that's another discussion).

**Comment by ngld** [comment]
It's nice to be back. ^^

I think the `LogLine` event dates back to ACTWS and changing it would break overlays like MopiMopi which rely on it. Simplifying that event wasn't worth breaking overlays or convincing the overlay devs to switch to OverlayPlugin's new events instead.
